nixos/modules/system/activation/switch-to-configuration.py

```diff
@@ -0,0 +1,611 @@
+    print_err("setting up tmpfiles")
+    if subprocess.call(["@systemd@/bin/systemd-tmpfiles", "--create", "--remove",
+                        "--exclude-prefix=/dev"]) != 0:
+        res = 3
+
+    # Reload units that need it. This includes remounting changed mount
+    # units.
+    if len(transitions.units_to_reload()) > 0:
+        units = sorted(transitions.units_to_reload())
+        print_err("reloading the following units: ", ", ".join(units))
+        if subprocess.call(["@systemd@/bin/systemctl", "reload", "--"] +
+                           cast(List[str], units)) != 0:
+            res = 4
+        os.unlink(RELOAD_UNIT_FILE)
+
+    # Restart changed services (those that have to be restarted rather
+    # than stopped and started).
+    if len(transitions.units_to_restart()) > 0:
+        units = sorted(transitions.units_to_restart())
+        print_err("restarting the following units: ", ", ".join(units))
+        if subprocess.call(["@systemd@/bin/systemctl", "restart", "--"] +
+                           cast(List[str], units)) != 0:
+            res = 4
+        os.unlink(RESTART_UNIT_FILE)
+
+    # Start all active targets, as well as changed units we stopped above.
+    # The latter is necessary because some may not be dependencies of the
+    # targets (i.e., they were manually started).  FIXME: detect units
+    # that are symlinks to other units.  We shouldn't start both at the
+    # same time because we'll get a "Failed to add path to set" error from
+    # systemd.
+    if len(transitions.filtered_units_to_start()) > 0:
+        print("starting the following units: ",
+            ", ".join(transitions.filtered_units_to_start()))
+    if subprocess.call(["@systemd@/bin/systemctl", "start", "--"] +
+                        sorted(transitions.units_to_start())) != 0:
+        res = 4
+    os.unlink(START_UNIT_FILE)
+
+    # Print failed and new units.
+    failed = []
+    new = []
+    active_new = get_active_units()
+    for unit, state in active_new.items():
+        if state.state == 'failed':
+            failed.append(unit)
+        elif state.state == 'auto-restart':
+            # A unit in auto-restart state is a failure *if* it previously failed to start
+            lines = subprocess.check_output(['systemd@/bin/systemctl', 'show', unit])
+            info = parse_key_values(lines.split('\n'))
+
+            if info['ExecMainStatus'] != '0':
+                failed.append(unit)
+        elif state.state != 'failed' and unit not in active_prev:
+            new.append(unit)
+
+    if len(new) > 0:
+        print_err("the following new units were started: ", ", ".join(sorted(new)))
+
+    if len(failed) > 0:
+        print_err("warning: the following units failed: ", ", ".join(sorted(failed)))
+        for unit in failed:
+            print_err('')
+            cmd = "COLUMNS=1000 @systemd@/bin/systemctl status --no-pager '{unit}' >&2".format(unit=unit)
+            subprocess.check_call(cmd, shell=True)
```

Comment:
When displaying error message you probably don't want to fail half-way: 
```diff
- subprocess.check_call(cmd, shell=True) 
+ subprocess.call(cmd, shell=True) 
```

Reply:
It should be also possible to avoid the shell:

```
subprocess.run(["@systemd@/bin/systemctl", "status", "--no-pager", unit], env=dict(COLUMNS="1000"), stdout=sys.stderr)
```